app/forms/embargo_form.rb

```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+class EmbargoForm
+  include ActiveModel::Model
+
+  attr_reader :work
+
+  attr_writer :embargoed_until
+  attr_accessor :remove
+
+  validate :embargoed_until_is_valid_date,
+           unless: :remove?
+
+  def initialize(work:, params:)
+    @work = work
+    super(params)
+  end
+
+  def embargoed_until
+    work_embargoed_until = if work.embargoed_until.respond_to?(:strftime)
+                             work.embargoed_until.strftime('%Y-%m-%d')
+                           end
+    @embargoed_until || work_embargoed_until
+  end
+
+  def remove?
+    !!ActiveModel::Type::Boolean.new.cast(remove)
+  end
+
+  def save
+    return false unless valid?
+
+    work.embargoed_until = if remove?
+                             nil
+                           elsif embargoed_until.present?
+                             Time.zone.parse(embargoed_until).beginning_of_day
+                           end
+    work.save
+  end
+
+  private
+
+    def embargoed_until_is_valid_date
+      return if embargoed_until.blank?
+
+      valid_format = embargoed_until.to_s.match(/^\d{4}-\d{2}-\d{2}$/)
+
+      valid_date = begin
+                     Date.parse(embargoed_until.to_s)
+                     true
+                   rescue ArgumentError, TypeError
+                     false
+                   end
+
+      if !valid_format
+        errors.add(:embargoed_until, :format)
+        return
+      end
+
+      if !valid_date
+        errors.add(:embargoed_until, :not_a_date)
+        nil
+      end
```

Comment:
What about:
``` ruby
def embargoed_until_is_valid_date
  return if embargoed_until.blank?
  
  unless embargoed_until.to_s.match(/^\d{4}-\d{2}-\d{2}$/)
    errors.add(:embargoed_until, :format)
    return
  end

  begin
    Date.parse(embargoed_until.to_s)
  rescue ArgumentError, TypeError
    errors.add(:embargoed_until, :not_a_date)
  end
end
```
I'm not sure if it's any clearer. I was thinking about suggesting _two_ new methods, one for format and one for date, and then `embargoed_until_is_valid_date` calls the two and returns early if the format's wrong.

It's not clear to me what the method is supposed to return, or if that's even important, but I think the only this it really needs to do is add errors or not.

Reply:
Yeah, this method doesn't need to return anything, just call `errors.add` if necessary. On the one hand, I sort of like to "set up all the stuff" in one block at the top of a method, then perform actions on it in all in the second half of a method, but your version is quite a bit cleaner. Maybe I'll go with that